indexing-hadoop/src/main/java/io/druid/indexer/InputRowHelper.java

```diff
@@ -0,0 +1,259 @@
+      sw.readFields(in);
+      List<String> dimensions = Arrays.asList(sw.toStrings());
+
+      long timestamp = in.readLong();
+
+      MapWritable mw = new MapWritable();
+      mw.readFields(in);
+
+      Map<String, Object> event = Maps.newHashMap();
+
+      //Read dimensions
+      for (String d : dimensions) {
+        Writable v = mw.get(new Text(d));
+
+        if (v == null) {
+          continue;
+        }
+
+        if (v instanceof Text) {
+          event.put(d, ((Text) v).toString());
+        } else if (v instanceof StringArrayWritable) {
+          event.put(d, Arrays.asList(((StringArrayWritable) v).toStrings()));
+        } else {
+          throw new ISE("unknown dim value type %s", v.getClass().getName());
+        }
+      }
+
+      //Read metrics
+      for (AggregatorFactory aggFactory : aggs) {
+        String k = aggFactory.getName();
+        Writable v = mw.get(new Text(k));
+
+        if (v == null) {
+          continue;
+        }
+
+        String t = aggFactory.getTypeName();
+
+        if (t.equals("float")) {
+          event.put(k, ((FloatWritable) v).get());
+        } else if (t.equals("long")) {
+          event.put(k, ((LongWritable) v).get());
+        } else {
+          //its a complex metric
+          ComplexMetricSerde serde = ComplexMetrics.getSerdeForType(t);
+          if (serde == null) {
+            throw new ISE("Unknown type[%s]", t);
+          }
+          BytesWritable bw = (BytesWritable) v;
+          event.put(k, serde.getObjectStrategy().fromByteBuffer(ByteBuffer.wrap(bw.getBytes()), bw.getLength()));
+        }
+      }
+
+      return new MapBasedInputRow(timestamp, dimensions, event);
+    } catch(IOException ex) {
+      throw Throwables.propagate(ex);
+    }
+  }
+
+  public final static InputRow getInputRowFromRow(final Row row, final List<String> dimensions) {
+    return new InputRow()
+    {
+      @Override
+      public List<String> getDimensions()
+      {
+        return dimensions;
+      }
+
+      @Override
+      public long getTimestampFromEpoch()
+      {
+        return row.getTimestampFromEpoch();
+      }
+
+      @Override
+      public DateTime getTimestamp()
+      {
+        return row.getTimestamp();
+      }
+
+      @Override
+      public List<String> getDimension(String dimension)
+      {
+        return row.getDimension(dimension);
+      }
+
+      @Override
+      public Object getRaw(String dimension)
+      {
+        return row.getRaw(dimension);
+      }
+
+      @Override
+      public float getFloatMetric(String metric)
+      {
+        return row.getFloatMetric(metric);
+      }
+
+      @Override
+      public long getLongMetric(String metric)
+      {
+        return row.getLongMetric(metric);
+      }
+
+      @Override
+      public int compareTo(Row o)
+      {
+        return row.compareTo(o);
+      }
+    };
+  }
+}
+
+class StringArrayWritable extends ArrayWritable
+{
+  public StringArrayWritable()
+  {
+    super(UTF8.class);
```

Comment:
true, but this is required if we use
ArrayWritable(String[] dims)
or else we would have to use ArrayWritable(Text.class, Text[] dims) which requires additional conversion on String[] dims.